src/ir_passes/lower/fsm.rs

```diff
@@ -1,17 +1,406 @@
-use super::{utils::interface_name, Binding, BuildCtx};
-use crate::ir;
+use std::{collections::HashMap, ops::Not};
+
+use super::{
+    utils::{cell_to_port_def, interface_name},
+    BuildCtx,
+};
+use crate::{ir, ir_passes::lower::utils::INTERFACE_PORTS};
+use calyx::{structure, Guard, Nothing};
 use calyx_ir::{self as calyx, RRC};
+use itertools::Itertools;
+
+#[derive(Default)]
+/// Represents an fsm component.
+pub(super) struct FsmBind {
+    /// Simple fsm components that have been created.
+    simple: HashMap<u64, calyx::Component>,
+    /// Raw counter fsms that have been created
+    counter: HashMap<u64, calyx::Component>,
+    /// Fsm components using `n` counters each with `d` states, stored under `(n, d)`.
+    counter_chain: HashMap<(u64, u64), calyx::Component>,
+}
+
+impl<'l> FsmBind {
+    /// Get an fsm with the number of states and minimum delay (II) from the binding
+    pub fn get(&'l mut self, states: u64, delay: u64) -> &'l calyx::Component {
+        self.add_opt(states, delay, None)
+    }
+
+    /// Insert an fsm with a number of states and a delay (II) into the binding
+    pub fn add(
+        &'l mut self,
+        states: u64,
+        delay: u64,
+        lib: &'l calyx::LibrarySignatures,
+    ) -> &'l calyx::Component {
+        self.add_opt(states, delay, Some(lib))
+    }
+
+    fn add_opt(
+        &'l mut self,
+        states: u64,
+        delay: u64,
+        lib: Option<&'l calyx::LibrarySignatures>,
+    ) -> &'l calyx::Component {
+        // TODO: Find a better metric to decide which type of fsm to generate.
```

Comment:
Let's create a tracking issue for this. We want to come up with some simple threshold to see when we can generate simple, shift-register-based FSMs vs the more complex counter ones. For extremely large delays, where the module is essentially sequential, we definitely want counter-based. On the other end of the spectrum, almost completely pipelined modules should probably use shift register based ones. We'll need to do some measurements to figure out which one to choose.

Reply:
Done at #273.

Reply:
Sounds good!